ab_tool/spreadsheets.py

```diff
@@ -59,25 +59,42 @@ def get_track_selection_csv(request, experiment, file_title="test.xlsx"):
     def row_generator():
         yield ["Student ID", "LIS Person Sourcedid", "Experiment", "Assigned Track"]
         # Write data to CSV file
-        for s in get_unassigned_students(request, experiment):
-            yield [s["student_id"], s["lis_person_sourcedid"], experiment.name, ""]
+        for student in get_unassigned_students(request, experiment):
+            yield [student, "", experiment.name, ""]
     return streamed_csv_response(row_generator(), file_title)
 
 
-def parse_uploaded_file(experiment, input_spreadsheet, filename):
-    users = {}
+def parse_uploaded_file(experiment, unassigned_students, input_spreadsheet, filename):
+    students = {}
+    errors = []
+    track_names = {track.name: track for track in experiment.tracks.all()}
     if filename.endswith('.csv'):
-        csvreader = csv.reader(input_spreadsheet.split("\n"))
-        for row in csvreader:
-            if row[2] == experiment:
-                users[row[0]] = row[3]
-        return users
+        # Slice off row 1 to skip headers
+        csvreader = csv.reader(input_spreadsheet.split("\n")[1:])
+        for row_number, row in enumerate(csvreader):
+            parse_row(row, row_number, experiment, track_names,
+                      unassigned_students, students, errors)
     elif filename.endswith('.xlsx') or filename.endswith('.xls'):
-        book = xlrd.open_workbook(file_contents=input_spreadsheet.read())
+        book = xlrd.open_workbook(file_contents=input_spreadsheet)
         sheet = book.sheet_by_index(0)
-        for row in range(sheet.nrows):
-            if sheet.cell_value(row, 2) == experiment:
-                users[sheet.cell_value(row, 0)] = sheet.cell_value(row, 3)
-        return users
+        # Start at 1 to skip headers
+        for row_number in range(1, sheet.nrows):
+            row = sheet.row_values(row_number)
+            parse_row(row, row_number, experiment, track_names,
+                      unassigned_students, students, errors)
     else:
         raise INVALID_FILE_TYPE
+    return students, errors
+
+
+def parse_row(row, row_number, experiment, tracks, unassigned_students, students, errors):
```

Comment:
This method is the extraction of a common code block of parsing csv files and xls files.  Any other way of cutting it would introduce several more lines of redundancy.  Comments added to this and parent function.